design/userdata-merging.md

```diff
@@ -0,0 +1,234 @@
+
+### Non-Goals
+
+ * Support other type of userdata, considering that Cluster-api only supports
+   cloud-init based bootstrap provider (CABPK) we will focus on cloud-init only
+   in this proposal.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+Considering a set of motley physical nodes, all with different networking
+setup, the user will be able to specify a per-node network configuration. Hence
+the BaremetalHost selection can be independent, not requiring to tie
+BaremetalHost and BaremetalMachines through labels to give the networking
+through a script passed to CABPK.
+
+#### Story 2
+
+Considering a CAPI deployment, it would be possible to specify some
+node-specific configuration related to the physical node, offering the user the
+possibility to fully use the deployment features such as scale-in/scale-out
+independently. The deployment could then use physical nodes that are not
+strictly identical.
+
+### Implementation Details/Notes/Constraints
+
+This does not require to modify the current CAPI workflow.
+When selecting a baremetalHost for the BaremetalMachine, the controller would
+merge the userdata from the machine with the user-provided input and the
+load-balancer related input. The output would be written as userdata in the
+BaremetalHost for deployment.
+
+### Risks and Mitigations
+
+There are several types of userData, such as cloud-init or ignition. This
+proposal should not implement a feature for one that would be detrimental or not
+consider another one.
+
+The user could also give some input that directly conflict with the output of
+the bootstrap provider. This needs to be handled carefully.
+
+## Design Details
+
+- A new field in the BaremetalHost spec would be added :
+
+```yaml
+  spec:
+    userDataInput:
+      type: cloud-init
+      userDataAppend: <base64 userdata file>
+      userDataPrepend: <base64 userdata file>
+```
+The type is compulsory, other fields are optional.
+
+- When merging the different inputs, comes the question of the order of actions
+  and the possible conflicts between the inputs. The user will be able to
+  specify how the userdata given will be merged with the output from CABPK. In
+  the case of cloud-init, the following applies.
+- For all fields that are lists, the content of the list, if it exists, in
+  userDataAppend will be appended to the content in this list from the
+  bootstrap provider. The content of the list, if it exists, in userDataPrepend
+  will be added at the beginning of the list from the bootstrap provider.
+- For maps, they will be merged and a duplicated key would trigger an error.
+- For other types, a duplicated key in the userdata would trigger an error too.
+- Not specifically to cloud-init, another place where conflict could occur is
+  in the files created by the bootstrap provider. If both the user input and the
+  bootstrap provider output are modifying the same file, an error should be
+  raised.
```

Comment:
So if I'm reading this correctly, the proposal to reimplement a cloud-config merger inside the cluster-api-provider, with merge behavior which differs from any of the current cloud-init merge strategies?